File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SecondRound/Chapter06/preorderTraversal_144.java
package algo.tzashinorpu.SecondRound.Chapter06;

import algo.tzashinorpu.SecondRound.TreeNode;

import java.util.LinkedList;
import java.util.List;

public class preorderTraversal_144 {
    public List<Integer> preorderTraversal(TreeNode root) {
        LinkedList<Integer> res = new LinkedList<>();
        dfs(root, res);
        return res;
    }

    private void dfs(TreeNode root, LinkedList<Integer> res) {
        if (root == null) {
            return;
        }
        res.add(root.val);
        dfs(root.left, res);
        dfs(root.right, res);
    }
}

File: Clion/labuladong/chapter0/binary_tree/preorderTraversal_144.cpp
//
// Created by tseringnorpu on 6/3/2023.
//
#include "../../header/header.h"

using namespace labuladong;

vector<int> Solution::preorderTraversal(TreeNode *root) {
    // root left right
    vector<int> res;
    if (!root) return res;
    res.push_back(root->val);
    auto t1 = preorderTraversal(root->left);
    res.insert(res.end(), t1.begin(), t1.end());
    auto t2 = preorderTraversal(root->right);
    res.insert(res.end(), t2.begin(), t2.end());
    return res;
}